compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt

```diff
@@ -0,0 +1,424 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.compose.ui.dnd
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.DisposableEffect
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.ExperimentalComposeUiApi
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.composed
+import androidx.compose.ui.geometry.Offset
+import androidx.compose.ui.graphics.painter.Painter
+import androidx.compose.ui.graphics.toPainter
+import androidx.compose.ui.layout.LayoutCoordinates
+import androidx.compose.ui.layout.boundsInWindow
+import androidx.compose.ui.layout.onGloballyPositioned
+import androidx.compose.ui.window.LocalWindow
+import androidx.compose.ui.window.density
+import java.awt.Image
+import java.awt.Point
+import java.awt.Window
+import java.awt.datatransfer.DataFlavor
+import java.awt.datatransfer.DataFlavor.selectBestTextFlavor
+import java.awt.datatransfer.Transferable
+import java.awt.dnd.DnDConstants
+import java.awt.dnd.DropTarget
+import java.awt.dnd.DropTargetDragEvent
+import java.awt.dnd.DropTargetDropEvent
+import java.awt.dnd.DropTargetEvent
+import java.awt.dnd.DropTargetListener
+import java.awt.image.BufferedImage
+import java.io.File
+
+/**
+ * Represent data types drag and dropped to an application from outside.
+ */
+@ExperimentalComposeUiApi
+sealed interface DropData {
+    /**
+     * Represents list of files drag and dropped to an application in a raw [java.net.URI] format.
+     */
+    data class FilesList(val rawUris: List<String>) : DropData
+
+    /**
+     * Represents an image drag and dropped to an application.
+     */
+    data class Image(val painter: Painter) : DropData
+
+    /**
+     * Represent text drag and dropped to an application.
+     *
+     * @param mimeType mimeType of the [content] such as "text/plain", "text/html", etc.
+     */
+    data class Text(val content: String, val mimeType: String?) : DropData
+}
+
+/**
+ * Adds detector of external drag and drop (e.g. files DnD from Finder to an application)
+ *
+ * @param onDragStart will be called when the pointer with external content entered the component.
+ * @param onDrag will be called for all drag events inside the component.
+ * @param onDrop is called when the pointer is released with [DropData] the pointer held.
+ * @param onDragCancel is called if the pointer exited the component bounds or unknown data was dropped.
+ */
+@ExperimentalComposeUiApi
+@Composable
+fun Modifier.onExternalDrag(
+    enabled: Boolean = true,
+    onDragStart: (Offset) -> Unit = {},
```

Comment:
Can we extend API, so we can know what mime type is dragging right now? I.e. we can pass `DropData` here, if we change its interface as described above.

Reply:
Not sure, will take a look at it

Reply:
Checked, we can, will do so